auth/examples_test.go

```diff
@@ -18,24 +21,63 @@ func userClaimFromToken(struct{}) string {
 	return "foobar"
 }
 
-// Simple example of server initialization code.
-func Example_serverConfig() {
-	exampleAuthFunc := func(ctx context.Context) (context.Context, error) {
-		token, err := grpc_auth.AuthFromMD(ctx, "bearer")
-		if err != nil {
-			return nil, err
-		}
-		tokenInfo, err := parseToken(token)
-		if err != nil {
-			return nil, status.Errorf(codes.Unauthenticated, "invalid auth token: %v", err)
-		}
-		grpc_ctxtags.Extract(ctx).Set("auth.sub", userClaimFromToken(tokenInfo))
-		newCtx := context.WithValue(ctx, "tokenInfo", tokenInfo)
-		return newCtx, nil
+//exampleAuthFunc is used by a middleware to authenticate requests
+func exampleAuthFunc(ctx context.Context) (context.Context, error) {
+	token, err := grpc_auth.AuthFromMD(ctx, "bearer")
+	if err != nil {
+		return nil, err
+	}
+
+	tokenInfo, err := parseToken(token)
+	if err != nil {
+		return nil, status.Errorf(codes.Unauthenticated, "invalid auth token: %v", err)
 	}
 
+	grpc_ctxtags.Extract(ctx).Set("auth.sub", userClaimFromToken(tokenInfo))
+
+	//WARNING: in production define your own type to avoid context collisions
+	newCtx := context.WithValue(ctx, "tokenInfo", tokenInfo)
+
+	return newCtx, nil
+}
+
+// Simple example of server initialization code
+func Example_serverConfig() {
 	_ = grpc.NewServer(
 		grpc.StreamInterceptor(grpc_auth.StreamServerInterceptor(exampleAuthFunc)),
 		grpc.UnaryInterceptor(grpc_auth.UnaryServerInterceptor(exampleAuthFunc)),
 	)
 }
+
+type gRPCserverAuthenticated struct{}
+
+//SayHello only can be called by client when authenticated by exampleAuthFunc
+func (g gRPCserverAuthenticated) SayHello(ctx context.Context, request *pb.HelloRequest) (*pb.HelloReply, error) {
+	return &pb.HelloReply{Message: "pong authenticated"}, nil
+}
+
+type gRPCserverUnauthenticated struct{}
+
+//SayHello can be called by client without being authenticated by exampleAuthFunc as AuthFuncOverride is called instead
+func (g *gRPCserverUnauthenticated) SayHello(ctx context.Context, request *pb.HelloRequest) (*pb.HelloReply, error) {
+	return &pb.HelloReply{Message: "pong unauthenticated"}, nil
+}
+
+//AuthFuncOverride is called instead of exampleAuthFunc
+func (g *gRPCserverUnauthenticated) AuthFuncOverride(ctx context.Context, fullMethodName string) (context.Context, error) {
+	log.Println("client is calling method:", fullMethodName)
+	return ctx, nil
+}
+
+// Simple example of server initialization code with AuthFuncOverride method.
+func Example_serverConfigWithAuthOverride() {
+	server := grpc.NewServer(
+		grpc.StreamInterceptor(grpc_auth.StreamServerInterceptor(exampleAuthFunc)),
+		grpc.UnaryInterceptor(grpc_auth.UnaryServerInterceptor(exampleAuthFunc)),
+	)
+
+	//in this case a single service is used,
+	//in most use cased there would be at least two services defined in the proto file
+	pb.RegisterGreeterServer(server, &gRPCserverAuthenticated{})
+	pb.RegisterGreeterServer(server, &gRPCserverUnauthenticated{})
```

Comment:
Registering both doesn't make sense, so how about we have a boolean or something to show the toggle between authenticating with and without the override?
```suggestion
	if useCustomAuth {
		pb.RegisterGreeterServer(server, &gRPCserverUnauthenticated{})
	else {
		pb.RegisterGreeterServer(server, &gRPCserverAuthenticated{})
	}
```

Reply:
true